### PR TITLE
feat(gatsby-transformer-react-docgen): allow parsing typescript files

### PR DESCRIPTION
**Pull request by jquense**
Typescript has no mime type :/

### CODE REVIEW

File: packages/gatsby-transformer-react-docgen/src/on-node-create.js [hunk 1]
```diff
@@ -10,6 +10,19 @@ const digest = str =>
 const propsId = (parentId, name) => `${parentId}--ComponentProp-${name}`
 const descId = parentId => `${parentId}--ComponentDescription`
 
+function canParse(node) {
+  return (
+    node &&
```
Comment (DSchau) [inline packages/gatsby-transformer-react-docgen/src/on-node-create.js]: I don't think this check is really needed!

```suggestion
```

File: packages/gatsby-transformer-react-docgen/src/on-node-create.js [hunk 2]
```diff
@@ -10,6 +10,19 @@ const digest = str =>
...
+    (node.internal.mediaType === `application/javascript` ||
+      // Typescript doesn't really have a mime type and .ts files are a media file :/
+      node.internal.mediaType === `application/typescript` ||
```
Comment (DSchau) [inline packages/gatsby-transformer-react-docgen/src/on-node-create.js]: Is there a reason we need _all_ of these checks? I would've thought there's one condition we could reliably use, i.e.

```js
return (
  node.internal.mediaType === `application/javascript` ||
  node.internal.mediaType === `application/typescript`
)
```

File: packages/gatsby-transformer-react-docgen/src/on-node-create.js (hunk 1 above)
Comment (jquense) [reply packages/gatsby-transformer-react-docgen/src/on-node-create.js]: ya it more out of an abundance of caution i can remove if you like?

Comment (DSchau) [reply packages/gatsby-transformer-react-docgen/src/on-node-create.js]: It's probably OK. Let's just merge this, thank you!

**Comment by DSchau** [comment]
Published as [`gatsby-transformer-react-docgen@3.0.3`](https://github.com/gatsbyjs/gatsby/blob/8830d47dd72e049e007ffeaeb966ec1632afd857/packages/gatsby-transformer-react-docgen/CHANGELOG.md#303-2019-01-29)

**Comment by jquense** [comment]
🎉

**Comment by thchia** [comment]
@jquense I'm getting the following error when attempting to parse a TS file:

```bash
error There was a problem parsing component metadata for file: "layout.tsx"

   8 | interface LayoutProps {
   9 |   children?: React.ReactNode
> 10 |   foo: string
     |  ^
  11 | }
  12 |
  13 | /**


  SyntaxError: Unexpected token (10:2)

- index.js:3834 Object.raise
    [gatsby-ts-docgen]/[@babel]/parser/lib/index.js:3834:17

  - index.js:5142 Object.unexpected
    [gatsby-ts-docgen]/[@babel]/parser/lib/index.js:5142:16

  - index.js:1542 Object.flowObjectTypeSemicolon
    [gatsby-ts-docgen]/[@babel]/parser/lib/index.js:1542:12

...
```

Seems to be related to the `flowObjectTypeSemicolon` check during the parse operation. AFAIK, omitting the semicolon is valid TS.

**Comment by jquense** [comment]
@thchia yeah at the moment full ts support isn't there, this adds the ability to even see .ts/.tsx files. We still need better upstream react-docgen support, e.g. https://github.com/reactjs/react-docgen/pull/320

**Comment by thchia** [comment]
@jquense would adding https://github.com/styleguidist/react-docgen-typescript be of any help? I'm playing around now trying to switch parsers based on the file type.

**Comment by jquense** [comment]
I'd really like to _not_ deal with that, and instead just snag the proptypes handler from there and use it with the original react-docgen.

**Comment by thchia** [comment]
Do you mean not deal with adding an extra dependency? Because it looks like the snaggable bits form the bulk of the code anyway. Either way I'm with you on not having multiple parser instances.

**Comment by jquense** [comment]
it's a few things, not having the extra dep but also not having to deal with inconsistencies and differences between the two tools.

**Pull request merged**